compiler/p/codegen/ControlFlowEvaluator.cpp

```diff
@@ -1693,864 +2355,1027 @@ TR::Register *OMR::Power::TreeEvaluator::ifacmpneEvaluator(TR::Node *node, TR::C
+ *
+ * \param secondChild
+ *    The node corresponding to the right-hand side operand of the comparison.
+ *
+ * \param compareInfo
+ *    The CompareInfo struct containing information about the comparison.
+ *
+ * \param cg
+ *    The code generator.
+ *
+ * \return
+ *    true if the comparison is a signed comparison, does not use register pairs, and is against
+ *    a value with a known sign bit; false otherwise.
+ */
+bool isSimpleSignedCompareToKnownSign(TR::Node *secondChild, const CompareInfo& compareInfo, TR::CodeGenerator *cg)
+   {
+   if (compareInfo.isUnsigned)
+      return false;
 
-   node->setRegister(trgReg);
-   cg->decReferenceCount(firstChild);
-   cg->decReferenceCount(secondChild);
-   return trgReg;
+   if (compareInfo.type == TR::Int64 && !cg->comp()->target().is64Bit())
+      return false;
+
+   return secondChild->isNonNegative() || secondChild->isNegative() || secondChild->getOpCode().isLoadConst();
    }
 
-void OMR::Power::TreeEvaluator::genBranchSequence(TR::Node* node,
-   TR::Register* src1Reg, TR::Register* src2Reg, TR::Register* trgReg,
-   TR::InstOpCode::Mnemonic branchOpCode, TR::InstOpCode::Mnemonic cmpOpCode, TR::CodeGenerator* cg)
+/**
+ * \brief
+ *    Gets the length of an integral type in bits.
+ *
+ * \param dt
+ *    The data type to get the length of.
+ *
+ * \param cg
+ *    The code generator.
+ *
+ * \return
+ *    The length of the provided type in bits.
+ */
+int32_t getTypeBitLength(TR::DataTypes dt, TR::CodeGenerator *cg)
    {
-   TR::Register    *condReg = cg->allocateRegister(TR_CCR);
-   TR::PPCControlFlowInstruction *cfop = (TR::PPCControlFlowInstruction *)
-      generateControlFlowInstruction(cg, TR::InstOpCode::setbool, node);
-   cfop->addTargetRegister(condReg);
-   cfop->addTargetRegister(trgReg);
-   cfop->addSourceRegister(src1Reg);
-   cfop->addSourceRegister(src2Reg);
-   cfop->setOpCode2Value(branchOpCode);
-   cfop->setCmpOpValue(cmpOpCode);
-   cg->stopUsingRegister(condReg);
+   switch (dt)
+      {
+      case TR::Int8:
+         return 8;
+      case TR::Int16:
+         return 16;
+      case TR::Int32:
+         return 32;
+      case TR::Int64:
+         return 64;
+      case TR::Address:
+         return cg->comp()->target().is64Bit() ? 64 : 32;
+      default:
+         TR_ASSERT_FATAL(false, "Invalid data type %s for getTypeBitLength", TR::DataType::getName(dt));
+      }
    }
 
+/**
+ * \brief
+ *    Determines whether an integral type is smaller than the size of a GPR on the current
+ *    machine.
+ *
+ * \param dt
+ *    The data type to check.
+ *
+ * \param cg
+ *    The code generator.
+ *
+ * \return
+ *    true if the provided type is smaller than the size of a GPR; false otherwise.
+ */
+bool isTypeSubRegister(TR::DataTypes dt, TR::CodeGenerator *cg)
+   {
+   switch (dt)
+      {
+      case TR::Int8:
+      case TR::Int16:
+         return true;
+      case TR::Int32:
+         return cg->comp()->target().is64Bit();
```

Comment:
similarly, is32Bit() only means the mode.  physically, 32bit value is always sub-register type by assuming support 64bit hw only.